crates/turborepo-paths/src/absolute_system_path.rs

```diff
@@ -406,9 +406,12 @@ impl AbsoluteSystemPath {
     /// Attempts to read a file, and:
     /// If the file does not exist it returns the default value.
     /// For all other scenarios passes through the `read_to_string` results.
-    pub fn read_or_default(&self, default: String) -> Result<String, io::Error> {
+    pub fn read_or_default<I>(&self, default_value: I) -> Result<String, io::Error>
```

Comment:
This is in no way blocking and if there is a change it can easily happen down the road.

I still think what we want is `pub fn read_optional_string(&self) -> Result<Option<String>, io::Error>`* as it gives the caller insight into what path the call went down. It gives greater flexibility for future use where users might have a default that would be better expressed as a non-`String` e.g.
```
let config_contents = config_path.read_optional_string()?;
let config: MyConfig = config_contents.map(|s| serde_json::from_str(&s)).unwrap_or_else(|| Ok(MyConfig { lever: true, knob: 12 })?
```

*And to be honest I feel that it should be named `read_string(&self) -> Result<Option<String>, io::Error>` and the current `read_string` should get renamed to `read_required_string`

If we want to keep the current behavior. Then there are two changes that could be made to keep with Rust conventions, I believe that this should get changed to:
```
pub fn read_or_else(&self, f: impl FnOnce() -> String) -> Result<String, io::Error> 
```
- The change to `or_else` over `or_default` matches the stdlib convention where `or_default` means that `T::default()` is used in the error case e.g. [`Option::unwrap_or_default`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_default)
- The change to taking a closure instead of an `Into<String>` gives greater flexibility to the caller where they might want to have a dynamic default: e.g. `config_path.read_or_else(|| format!("default_value: {foo}"))`

Reply:
If we surface this as a `FnOnce` via something like `read_to_string_or_else` it's purpose-defeating. That mapping ends up looking exactly like the body of the helper function, and copied straight into a dozen places in the code base.

```rust
path
    .read_to_string()
    .or_else(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            Ok("{}".into())
        } else {
            Err(e)
        }
    })
```

The other alternative signature that you're proposing, usage looks like this:

```rust
// Existing in this PR:
pub fn read_or_fallback(&self, default: Into<String>) -> Result<String, io:Error> {}
let contents = path.read_or_fallback("{}")?;

// Proposed:
pub fn read_existing_to_string(&self) -> Result<Option<String>, io:Error> {}

// Allocates if you're not careful.
let contents = path
    .read_existing_to_string()?
    .unwrap_or("{}".to_string());

// No allocation but a _lot_ of ceremony.
let contents = path
    .read_existing_to_string()?
    .unwrap_or_else(|| "{}".to_string());
```

I don't hate it, but it's not the same pit-of-success API. To make the method name look like Rust I've done this:
```rust
pub fn read_existing_to_string_or(&self, default: Option<Into<String>>) -> Result<String, io:Error> {}
let contents = path.read_existing_to_string_or(Ok("{}"))?;
```

Reply:
Assuming we had:
```rust
pub fn read_existing_to_string(&self) -> Result<Option<String>, io:Error> {}
```

How would you propose reducing the boilerplate across invocations at multiple files?
```rust
let contents = path
    .read_existing_to_string()?
    .unwrap_or_else(|| "{}".to_string());
```

I just want a nice-to-use API (`let contents = path.read_or_default("{}")?;`) at the callsite. Is there a less-verbose, more-Rust-like API for this that I'm just not familiar with? Is this a `base64` type of situation?

https://twitter.com/mycoliza/status/1708933310548935148

Reply:
> I don't hate it, but it's not the same pit-of-success AP

Clippy is already good at yelling at you if your `unwrap_or` contains a function call: https://rust-lang.github.io/rust-clippy/v0.0.212/index.html#or_fun_call

> How would you propose reducing the boilerplate across invocations at multiple files?

```
pub fn read_json(path: &AbsoluteSystemPath) -> Result<String>, io::Error> {
    path
        .read_existing_to_string()?
        .unwrap_or_else(|| "{}".to_string())
}
```
Not wonderful, but it allows the API to be flexible for future callers which might not want to provide a default in the form of a `String`.

> Is this a base64 type of situation?

To a degree, yes. I've drunk the Rust kool-aid and I don't find `foo.unwrap_or_else(|| "{}".to_string())` to be an eyesore/verbose. I'm used to the Rust `std` and I tend to try to model most APIs to pertain to the conventions laid out by it. Upshot of following those conventions is that tools like `clippy` are built around proper usage of the `std` APIs and we can rely on `clippy` telling callers how to use our library correctly.